avatars/ragdoll.js

```diff
@@ -0,0 +1,507 @@
+          return Math.max(0.1, boneRadius * 2);
+        } else {
+          if (children.length === 0) {
+            const diff = new THREE.Vector3().setFromMatrixPosition(modelBone.matrixWorld)
+              .sub(new THREE.Vector3().setFromMatrixPosition(modelBone.parent.matrixWorld));
+            const length = diff.length();
+            modelBoneEnd = modelBoneStart.clone()
+              .add(
+                diff,
+              );
+            return Math.max(0.08, length); // same length as parent
+          } else {
+            // todo: use real bone end, instead of calculated.
+            const acc = new THREE.Vector3();
+            for (const child of children) {
+              localVector.setFromMatrixPosition(child.matrixWorld);
+              acc.add(localVector);
+            }
+            modelBoneEnd = acc.divideScalar(children.length);
+            return Math.max(0.08, modelBoneStart.distanceTo(modelBoneEnd));
+          }
+        }
+      })();
+      modelBone.boneLength = boneLength;
+      modelBone.modelBoneEnd = modelBone.worldToLocal(modelBoneEnd.clone());
+      meshBone.boneLength = boneLength;
+      meshBone.modelBoneEnd = modelBone.modelBoneEnd;
+
+      // forward quaternion
+      if (k === 'Hips') {
+        modelBone.forwardQuaternion = new THREE.Quaternion()
+          .setFromAxisAngle(new THREE.Vector3(1, 0, 0), -Math.PI / 2);
+      } else {
+        modelBone.forwardQuaternion = new THREE.Quaternion()
+          .setFromRotationMatrix(
+            localMatrix.lookAt(
+              modelBoneStart,
+              modelBoneEnd,
+              localVector2.set(0, 1, 0),
+            ),
+          );
+      }
+
+      // set capsule geometries
+      meshBone.physicsMesh.geometry = _makeCapsuleGeometry(meshBone);
+
+      // memoize
+      // modelBoneToFlatMeshBoneMap.set(modelBone, meshBone);
+    }
+  };
+  object.createRagdoll = avatar => {
+    if (object.isCreatedRagdoll) return;
+    object.isCreatedRagdoll = true;
+    for (const k in flatMeshes) {
+      const meshBone = flatMeshes[k];
+      // const body = physx.physxWorker.addBoxGeometryPhysics(physx.physics, meshBone.position, meshBone.quaternion, meshBone.sizeHalf, meshBone.physicsId, true, localPlayer.characterController.physicsId); // can't hardcode localPlayer here, used disableGeometryQueries instead.
+      const body = physx.physxWorker.addBoxGeometryPhysics(physx.physics, meshBone.position, meshBone.quaternion, meshBone.sizeHalf, meshBone.physicsId, true);
+      physicsManager.disableGeometryQueries(meshBone); // prevent own CCT collide with ragdoll bones.
+      avatar.app.physicsObjects.push(meshBone);
+      // console.log('mass 1: ', physicsManager.getBodyMass(body));
+      // physicsManager.updateMassAndInertia(body, 0.000001);
+      physicsManager.updateMassAndInertia(body, 0); // note: set mass 0 ( ie kinematic? ) will not break joints and get good result, but got much slow animation.
```

Comment:
I think I understand this code better now. Can we _not_ return the body from addBoxGeometryPhysics? It seems inconsistent with the rest of the code.

We are already injecting the physics ID which I think can be used to look up the body on the WASM side, so there is no need to add yet another id being returned here.

Then, `updateMassAndInertia` can just take the physics ID.

Reply:
Changed. https://github.com/webaverse/app/pull/2731/commits/02f681682908ef3913316bbe8b1ac0202b170170
Sibling PR: https://github.com/webaverse/app-wasm/pull/44